lib/spack/spack/cmd/container.py

```diff
@@ -0,0 +1,151 @@
+        help='do not use checksums to verify downloaded files (unsafe)')
+    subparser.add_argument(
+        '--show-log-on-error', action='store_true', default=False,
+        help='print full build log if build fails')
+    subparser.add_argument(
+        '--spack-branch', '--branch', action='store', type=str,
+        default='master',
+        help='branch of Spack to deploy (default: master)')
+
+
+def container(parser, args):
+    if not args.specs:
+        tty.die("spack spec requires at least one spec")
+
+    recipe = hpccm.Stage()
+
+    recipe += baseimage(image=args.baseimage, _distro=args.distro)
+
+    # OS dependencies
+    recipe += hpccm.building_blocks.packages(
+        apt=['autoconf', 'build-essential', 'bzip2', 'ca-certificates',
+             'coreutils', 'cpio', 'curl', 'environment-modules', 'git',
+             'gzip', 'libssl-dev', 'make', 'openssh-client', 'patch',
+             'pkg-config', 'svn', 'tar', 'tcl', 'unzip', 'zlib1g'],
+        yum=['autoconf', 'bzip2', 'ca-certificates', 'coreutils', 'cpio',
+             'curl', 'environment-modules', 'git', 'gzip', 'make',
+             'openssh-clients', 'openssl-devel', 'patch', 'pkg-config',
+             'svn', 'tar', 'tcl', 'unzip', 'zlib-devel'])
+
+    # GNU compilers
+    recipe += hpccm.building_blocks.gnu()
+
+    # Spack itself
+    spack_bb = bb_base()
+    spack_bb += comment('Spack')
+    spack_bb += shell(commands=[
+        'mkdir -p /opt && cd /opt',
+        'git clone --depth=1 --branch {} https://github.com/spack/spack'.format(args.spack_branch),
+        '/opt/spack/bin/spack bootstrap',
+        'ln -s /opt/spack/share/spack/setup-env.sh /etc/profile.d/spack.sh',
+        'ln -s /opt/spack/share/spack/spack-completion.bash /etc/profile.d'])
+    spack_bb += environment(variables={'PATH': '/opt/spack/bin:$PATH',
+                                       'FORCE_UNSAFE_CONFIGURE': '1'})
+    recipe += spack_bb
+
+    # spack install command line arguments
+    install_args = ['-y']
+    if args.no_checksum:
+        install_args.append('--no-checksum')
+    if args.show_log_on_error:
+        install_args.append('--show-log-on-error')
+    install_args_str = ' '.join(install_args)
+
+    # Process each spec
+    for spec in spack.cmd.parse_specs(args.specs):
+        spec.concretize()
+
+        bb = bb_base()
+        bb += comment('{0} {1}'.format(spec.name, spec.version))
+
+        # Process each dependency.  Use a selectable depth to balance
+        # the build cache vs. container spec file complexity
+        for depth, dspec in spec.traverse_edges(order='post', depth=True):
+            if depth <= args.depth:
+                node = dspec.spec
+                cmd = 'spack install {0} {1}@{2}'.format(install_args_str,
+                                                         node.name,
+                                                         node.version)
+                bb += shell(chdir=False, commands=[cmd, 'spack clean --all'])
```

Comment:
I'm unfamiliar with hpccm and containers, does this `shell()` invocation correspond to a layer or container (or is `bb` the layer or container)?

Reply:
It's a layer.  It corresponds to a `RUN` instruction in a Dockerfile or a `%post%` section in a Singularity definition file. 

`bb` is a HPCCM building block - really just grouping to make the output slightly more human readable.